core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -380,29 +409,36 @@ DEFINE_UNARY_FUNCTION_EVAL(logb, 2);
-  struct MathBinaryFunction_##FUNC {                                      \
-    template <typename T, typename U>                                     \
-    static KOKKOS_FUNCTION auto eval(T x, U y) {                          \
-      static_assert(                                                      \
-          std::is_same<decltype(Kokkos::FUNC((T)0, (U)0)),                \
-                       math_binary_function_return_type_t<T, U>>::value); \
-      return Kokkos::FUNC(x, y);                                          \
-    }                                                                     \
-    template <typename T, typename U>                                     \
-    static auto eval_std(T x, U y) {                                      \
-      static_assert(                                                      \
-          std::is_same<decltype(std::FUNC((T)0, (U)0)),                   \
-                       math_binary_function_return_type_t<T, U>>::value); \
-      return std::FUNC(x, y);                                             \
-    }                                                                     \
-    static KOKKOS_FUNCTION double ulp_factor() { return ULP_FACTOR; }     \
-  };                                                                      \
-  using kk_##FUNC = MathBinaryFunction_##FUNC;                            \
-  template <>                                                             \
-  struct math_function_name<MathBinaryFunction_##FUNC> {                  \
-    static constexpr char name[] = #FUNC;                                 \
-  };                                                                      \
+#define DEFINE_BINARY_FUNCTION_EVAL(FUNC, ULP_FACTOR)                       \
+  struct MathBinaryFunction_##FUNC {                                        \
+    template <typename T, typename U>                                       \
+    static KOKKOS_FUNCTION auto eval(T x, U y) {                            \
+      static_assert(                                                        \
+          std::is_same<decltype(Kokkos::FUNC((T)0, (U)0)),                  \
+                       math_binary_function_return_type_t<T, U>>::value);   \
+      return Kokkos::FUNC(x, y);                                            \
+    }                                                                       \
+    template <typename T, typename U>                                       \
+    static auto eval_std(T x, U y) {                                        \
+      if constexpr (KE::kokkos_type_is_half_t<T>::value ||                  \
+                    KE::kokkos_type_is_bhalf_t<T>::value ||                 \
+                    KE::kokkos_type_is_half_t<U>::value ||                  \
+                    KE::kokkos_type_is_bhalf_t<U>::value) {                 \
+        return std::FUNC(static_cast<float>(x), static_cast<float>(y));     \
```

Comment:
Why are we casting all arguments instead of only those that are `half_t`?

Reply:
@dalg24, @masterleinad: Can we clean this up a bit with another wrapper using a logical-or:
```
template <class T>
struct kokkos_type_is_half : kokkos_type_is_half_t<T>::value || kokkos_type_is_bhalf_t<T>::value;
```
?